source/beamline/REIXS/REIXSSampleMotor.cpp

```diff
@@ -0,0 +1,377 @@
+{
+	AMPseudoMotorControl::onMoveSucceeded(action);
+	updateMinimumAndMaximum();
+}
+
+AMAction3 * REIXSSampleMotor::createMoveAction(double setpoint)
+{
+
+	AMListAction3* moveAction = new AMListAction3(new AMListActionInfo3(QString("In-plane sample move"),
+	                                                                   QString("In-place sample move")),
+				                                 AMListAction3::Parallel);
+
+
+	QVector3D targetGlobalVector = createGlobalMovementVector(setpoint);
+
+	moveAction->addSubAction(AMActionSupport::buildControlMoveAction(horizontalTranslationControl_,
+	                                                                 targetGlobalVector.x()));
+	moveAction->addSubAction(AMActionSupport::buildControlMoveAction(normalTranslationControl_,
+	                                                                 targetGlobalVector.y()));
+
+	return moveAction;
+}
+
+void REIXSSampleMotor::rotateSystem()
+{
+	if(isConnected()) {
+
+		QVector3D vX(1, 0, 0);
+		QVector3D vY(0, 1, 0);
+		QVector3D vZ(0, 0, 1);
+
+		QQuaternion qX = QQuaternion::fromAxisAndAngle(vX, 0);
+		QQuaternion qY = QQuaternion::fromAxisAndAngle(vY, 0);
+		QQuaternion qZ = QQuaternion::fromAxisAndAngle(vZ, -totalRotation());
+
+		rotationQuaternion_ = (qX * qY) * qZ;
+		rotationQuaternion_.normalize();
+
+		updateValue();
+		updateMinimumAndMaximum();
+	}
+
+}
+
+QVector3D REIXSSampleMotor::globalSystemToPrime(const QVector3D &globalVector) const
+{
+	return rotationQuaternion_.rotatedVector(globalVector);
+}
+
+QVector3D REIXSSampleMotor::primeSystemToGlobal(const QVector3D &primeVector) const
+{
+	return rotationQuaternion_.conjugate().rotatedVector(primeVector);
+}
+
+QVector3D REIXSSampleMotor::createGlobalMovementVector(double primeSetpoint) const
+{
+	// Get the current position in the global system
+	QVector3D startingGlobalVector(horizontalTranslationControl_->value(),
+	                               normalTranslationControl_->value(),
+	                               0);
+
+	// Translate that into the prime system
+	QVector3D startingPrimeVector = globalSystemToPrime(startingGlobalVector);
+
+	// Update the value to the setpoint depending on which axis we are.
+	if(direction_ == AMMotorGroupObject::HorizontalMotion) {
+
+		startingPrimeVector.setX(primeSetpoint);
+	} else if(direction_ == AMMotorGroupObject::NormalMotion ) {
+
+		startingPrimeVector.setY(primeSetpoint);
+	}
+
+	// Return the prime motion in terms of the global system
+	return primeSystemToGlobal(startingPrimeVector);
+}
+
+double REIXSSampleMotor::valueForDirection(const QVector3D &vector)
+{
+	switch (direction_) {
+
+	case AMMotorGroupObject::HorizontalMotion:
+		return vector.x();
+	case AMMotorGroupObject::VerticalMotion:
+		return vector.z();		// This shouldn't happen, but we include it anyway.
+	case AMMotorGroupObject::NormalMotion:
+		return vector.y();
+	}
+
+	return 0;
+}
+
+double REIXSSampleMotor::radiansToDegrees(double radians)
```

Comment:
There is a class in beamline/camera called AMAngle but it would bloat the code a little in here if I was to use it. To be honest I think this is a case where a namespace header file with a bunch of angle helper functions would be more useful than a class.